src/printer.js

```diff
@@ -109,6 +109,14 @@ function genericPrint(path, options, printPath) {
     needsParens = path.needsParens();
   }
 
+  if (node.type) {
+    // HACK: ASI prevention in no-semi mode relies on knowledge of whether
+    // or not a paren has been inserted (see `exprNeedsASIProtection()`).
+    // For now, we're just passing that information by mutating the AST here,
+    // but it would be nice to find a cleaner way to do this.
+    node.needsParens = needsParens;
+  }
```

Comment:
Here's my idea: I don't see any reason why we can't have access to the path instead of the node, and we can just call `path.needsParens()` multiple times. To fix the performance hit, we could cache that method so if it's already been calculated it just returns the value.

The only problem is that the path abstraction using mutation heavily, so writing the cache isn't too straight-forward, but it's doable. I think we've talked about making the path abstraction better anyway.

I'll make another comment below about this. But I don't think this code is complex enough to block this PR from landing. We can work on this later. I feel OK about letting this stuff in prettier because we aren't exposing these objects over an API or anything so these hacks don't impact users, as long as the output is good.